esp-hal/src/uart.rs

```diff
@@ -2651,6 +2487,91 @@ pub mod lp_uart {
     }
 }
 
+/// UART Peripheral Instance
+pub trait Instance: Peripheral<P = Self> + PeripheralMarker + Into<AnyUart> + 'static {
+    /// Returns the peripheral data describing this UART instance.
+    fn peripheral_data(&self) -> &UartPeripheral;
+}
+
+/// Peripheral data describing a particular UART instance.
+pub struct UartPeripheral {
+    /// Pointer to the register block for this UART instance.
+    ///
+    /// Use [Self::register_block] to access the register block.
+    pub register_block: *const RegisterBlock,
+
+    /// Interrupt handler for the asynchronous operations of this UART instance.
+    pub async_handler: InterruptHandler,
+
+    /// Waker for the asynchronous RX operations.
+    pub rx_waker: AtomicWaker,
```

Comment:
AtomicWaker forces these struct to live in RAM. We could split this into Peripheral and State instead, although I'm not sure if this is an issue.